// cdktf/lib/backends/cloud-backend.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaggedCloudWorkspaces = exports.NamedCloudWorkspace = exports.CloudWorkspace = exports.CloudBackend = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const util_1 = require("../util");
const remote_backend_1 = require("./remote-backend");
const terraform_functions_1 = require("../terraform-functions");
const terraform_backend_1 = require("../terraform-backend");
const validations_1 = require("../validations");
/**
 * The Cloud Backend synthesizes a {@link https://www.terraform.io/cli/cloud/settings#the-cloud-block cloud block}.
 * The cloud block is a nested block within the top-level terraform settings block.
 * It specifies which Terraform Cloud workspaces to use for the current working directory.
 * The cloud block only affects Terraform CLI's behavior.
 * When Terraform Cloud uses a configuration that contains a cloud block - for example, when a workspace is configured to use a VCS provider directly - it ignores the block and behaves according to its own workspace settings.
 */
class CloudBackend extends terraform_backend_1.TerraformBackend {
    constructor(scope, props) {
        super(scope, "backend", "cloud");
        this.props = props;
        this.node.addValidation(new validations_1.ValidateBinaryVersion("terraform", ">=1.1", `${util_1.terraformBinaryName} version`, `The cloud block is only supported for Terraform >=1.1. Please upgrade your Terraform version.`));
    }
    /**
     * Adds this resource to the terraform JSON output.
     */
    toTerraform() {
        return {
            terraform: {
                cloud: util_1.deepMerge(this.synthesizeAttributes(), this.rawOverrides),
            },
        };
    }
    synthesizeAttributes() {
        return util_1.keysToSnakeCase({
            ...this.props,
            workspaces: this.props.workspaces.toTerraform(),
        });
    }
    getRemoteStateDataSource(scope, name, _fromStack) {
        // The cloud "backend" does not have a coresponding remote state data source since it's meant to also
        // work with multiple workspaces through tags.
        if ("name" in this.props.workspaces) {
            return new remote_backend_1.DataTerraformRemoteState(scope, name, {
                organization: this.props.organization,
                workspaces: { name: this.props.workspaces.name },
                token: this.props.token,
                hostname: this.props.hostname,
            });
        }
        throw new Error("The Cloud backend only supports cross-stack references when the workspace is defined by name instead of by tags.");
    }
}
exports.CloudBackend = CloudBackend;
_a = JSII_RTTI_SYMBOL_1;
CloudBackend[_a] = { fqn: "cdktf.CloudBackend", version: "0.13.0" };
/**
 * A cloud workspace can either be a single named workspace, or a list of tagged workspaces.
 */
class CloudWorkspace {
}
exports.CloudWorkspace = CloudWorkspace;
_b = JSII_RTTI_SYMBOL_1;
CloudWorkspace[_b] = { fqn: "cdktf.CloudWorkspace", version: "0.13.0" };
/**
 * The name of a single Terraform Cloud workspace.
 * You will only be able to use the workspace specified in the configuration with this working directory, and cannot manage workspaces from the CLI (e.g. terraform workspace select or terraform workspace new).
 */
class NamedCloudWorkspace extends CloudWorkspace {
    constructor(name) {
        super();
        this.name = name;
    }
    toTerraform() {
        return {
            name: this.name,
        };
    }
}
exports.NamedCloudWorkspace = NamedCloudWorkspace;
_c = JSII_RTTI_SYMBOL_1;
NamedCloudWorkspace[_c] = { fqn: "cdktf.NamedCloudWorkspace", version: "0.13.0" };
/**
 *  A set of Terraform Cloud workspace tags. You will be able to use this working directory with any workspaces that have all of the specified tags, and can use the terraform workspace commands to switch between them or create new workspaces. New workspaces will automatically have the specified tags. This option conflicts with name.
 */
class TaggedCloudWorkspaces extends CloudWorkspace {
    constructor(tags) {
        super();
        this.tags = tags;
    }
    toTerraform() {
        return {
            tags: terraform_functions_1.Fn.toset(this.tags),
        };
    }
}
exports.TaggedCloudWorkspaces = TaggedCloudWorkspaces;
_d = JSII_RTTI_SYMBOL_1;
TaggedCloudWorkspaces[_d] = { fqn: "cdktf.TaggedCloudWorkspaces", version: "0.13.0" };
