// cdktf/lib/validations/validate-provider-presence.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidateProviderPresence = void 0;
// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
const constructs_1 = require("constructs");
const terraform_provider_1 = require("../terraform-provider");
const terraform_resource_1 = require("../terraform-resource");
const terraform_data_source_1 = require("../terraform-data-source");
/**
 * A validation that is added by default, ensuring that all providers
 * used are defined via Constructs.
 */
class ValidateProviderPresence {
    /**
     * Creates a ValidateProviderPresence
     */
    constructor(host) {
        this.host = host;
        /**
         * All provider names found in the construct tree.
         */
        this.providerNames = new Set();
        /**
         * All TerraformProviders found in the construct tree.
         */
        this.foundProviders = [];
    }
    /**
     * Checks the construct tree recursively
     */
    check(node) {
        if (node instanceof terraform_resource_1.TerraformResource ||
            node instanceof terraform_data_source_1.TerraformDataSource) {
            if (node.terraformGeneratorMetadata) {
                this.providerNames.add(node.terraformGeneratorMetadata.providerName);
            }
        }
        if (node instanceof terraform_provider_1.TerraformProvider) {
            this.foundProviders.push(node);
        }
        for (const child of constructs_1.Node.of(node).children) {
            this.check(child);
        }
    }
    /**
     * Run the validation
     */
    validate() {
        this.check(this.host);
        const missingProviders = Array.from(this.providerNames).filter((name) => {
            return !this.foundProviders.some((p) => p.terraformGeneratorMetadata?.providerName === name);
        });
        if (missingProviders.length === 0) {
            return [];
        }
        else {
            return [
                `Found resources without a matching provider construct. Please make sure to add provider constructs [e.g. new RandomProvider(...)] to your stack for the following providers: ${missingProviders.join(", ")}`,
            ];
        }
    }
}
exports.ValidateProviderPresence = ValidateProviderPresence;
